xwayland/selection/selection.c

```diff
@@ -32,16 +45,63 @@ void xwm_selection_transfer_destroy_property_reply(
 	transfer->property_reply = NULL;
 }
 
-void xwm_selection_transfer_init(struct wlr_xwm_selection_transfer *transfer) {
+void xwm_selection_transfer_init(struct wlr_xwm_selection_transfer *transfer,
+		struct wlr_xwm_selection *selection) {
+	struct wlr_xwm *xwm = selection->xwm;
+
+	transfer->selection = selection;
 	transfer->wl_client_fd = -1;
+
+	transfer->window = xcb_generate_id(xwm->xcb_conn);
+	xcb_create_window(
+		xwm->xcb_conn,
+		XCB_COPY_FROM_PARENT,
+		transfer->window,
+		xwm->screen->root,
+		0, 0,
+		10, 10,
+		0,
+		XCB_WINDOW_CLASS_INPUT_OUTPUT,
+		xwm->screen->root_visual,
+		XCB_CW_EVENT_MASK, (uint32_t[]){
+			XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY | XCB_EVENT_MASK_PROPERTY_CHANGE
+		}
+	);
+
+	xcb_set_selection_owner(xwm->xcb_conn, transfer->window,
+		xwm->atoms[CLIPBOARD_MANAGER], XCB_TIME_CURRENT_TIME);
+	xcb_flush(xwm->xcb_conn);
+}
+
+struct wlr_xwm_selection_transfer *xwm_selection_transfer_create_incoming(
+		struct wlr_xwm_selection *selection) {
+	struct wlr_xwm_selection_transfer *transfer = calloc(1, sizeof(*transfer));
+	if (!transfer) {
+		return NULL;
+	}
+
+	xwm_selection_transfer_init(transfer, selection);
+	wl_list_insert(&selection->incoming, &transfer->link);
+
+	return transfer;
 }
 
-void xwm_selection_transfer_finish(
+void xwm_selection_transfer_destroy(
 		struct wlr_xwm_selection_transfer *transfer) {
-	transfer->incr = false;
+	if (!transfer) {
+		return;
+	}
+
 	xwm_selection_transfer_destroy_property_reply(transfer);
 	xwm_selection_transfer_remove_event_source(transfer);
 	xwm_selection_transfer_close_wl_client_fd(transfer);
+
+	if (transfer->window) {
+		xcb_destroy_window(transfer->selection->xwm->xcb_conn, transfer->window);
```

Comment:
In theory we'd need to flush here. Maybe it's worth it to let the other side know we're not interested in the clipboard content anymore in case of a write error?